File: infer/tests/codetoanalyze/java/threadsafety/Alias.java
/*
 * Copyright (c) 2017 - present Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
import javax.annotation.concurrent.ThreadSafe;

@ThreadSafe
class Alias{
  A a,b;

  public void foo() {
    int x;
    synchronized(this){
      a = b;
      a.f = 101;
    }
    x = b.f;  // may_alias
  }

  public void bar(A a,A b) {
     int x;
     synchronized(this){
       a.f = 101;
     }
    x = b.f;  // no may_alias needed, argument treatment suffices
  }
}

class A { int f = 0; }

File: infer/tests/codetoanalyze/c/callgraph/alias.c
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <stdlib.h>

void set_both(int *p, int *q) {
  *p = 1;
  *q = 2;
}

void call_if_aliased(int *p, int *q) {
  if (p==q) set_both(p, q);
}

void always_call(int *p, int *q) {
  set_both(p, q);
}

void call_depending_on_alias(int *p, int *q) {
  call_if_aliased(p, q);
  always_call(p, q);
}

void test_aliased() {
  int x = 0;
  call_depending_on_alias(&x, &x);
}

void test_not_aliased() {
  int x = 0;
  int y = 1;
  call_depending_on_alias(&x, &y);
}
